src/sinks/file.rs

```diff
@@ -0,0 +1,563 @@
+                                file,
+                                BytesCodec::new(),
+                            ));
+                        }
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Err(err) => {
+                            self.state = FileSinkState::Disconnected;
+                            return Err(err);
+                        }
+                    },
+                }
+            }
+        }
+
+        fn encode_event(event: Event, encoding: &Option<Encoding>) -> Result<Bytes, ()> {
+            let log = event.into_log();
+
+            let result = match (encoding, log.is_structured()) {
+                (&Some(Encoding::Json), _) | (_, true) => serde_json::to_vec(&log.all_fields())
+                    .map_err(|e| panic!("Error encoding: {}", e)),
+
+                (&Some(Encoding::Text), _) | (_, false) => Ok(log
+                    .get(&event::MESSAGE)
+                    .map(|v| v.as_bytes().to_vec())
+                    .unwrap_or(Vec::new())),
+            };
+
+            result.map(|mut bytes| {
+                bytes.push(b'\n');
+                Bytes::from(bytes)
+            })
+        }
+    }
+
+    impl Sink for FileSink {
+        type SinkItem = Bytes;
+        type SinkError = io::Error;
+
+        fn start_send(
+            &mut self,
+            line: Self::SinkItem,
+        ) -> StartSend<Self::SinkItem, Self::SinkError> {
+            match self.poll_file() {
+                Ok(Async::Ready(file)) => {
+                    debug!(
+                        message = "sending event",
+                        bytes = &field::display(line.len())
+                    );
+                    match file.start_send(line) {
+                        Ok(ok) => Ok(ok),
+
+                        Err(err) => {
+                            self.state = FileSinkState::Disconnected;
+                            Err(err)
+                        }
+                    }
+                }
+                Ok(Async::NotReady) => Ok(AsyncSink::NotReady(line)),
+                Err(err) => Err(err),
+            }
+        }
+
+        fn poll_complete(&mut self) -> Result<Async<()>, Self::SinkError> {
+            if let FileSinkState::Disconnected = self.state {
+                return Err(disconnected());
+            }
+
+            let file = try_ready!(self.poll_file());
+
+            match file.poll_complete() {
+                Err(err) => {
+                    error!("Error while completing {:?}: {}", self.path, err);
+                    self.state = FileSinkState::Disconnected;
+                    Ok(Async::Ready(()))
+                }
+                Ok(ok) => Ok(ok),
+            }
+        }
+
+        fn close(&mut self) -> Poll<(), Self::SinkError> {
+            match self.poll_complete() {
+                Ok(Async::Ready(())) => match self.state {
```

Comment:
you can use `&mut self.state` here and avoid having to use `ref mut` below.

Reply:
thanks, it works